### ISSUE TITLE
Polimorphic adapters

### ISSUE THREAD

**Issue by sky-pedrosaraiva** [ngs-doo/dsl-json#258]
Hello, i'm trying to replace moshi with dsl-json.
In moshi is possible to implement adapters for a certain interface with multiple implementations:

```
PolymorphicJsonAdapterFactory.of(Dummy::class.java, "type")
        .withSubtype(SuperDummy::class.java, TypesEnum.SuperDummy.value)
        .withSubtype(MegaDummy::class.java, TypesEnum.MegaDummy.value)
```

This basically tells moshi to use the type field to map the correct implementation based on the value defined in a Enum.
I know that there is a descriminator but this doesnt seam to allow mapping a custom value.

Is this approach possible to implement in dsl-json?

**Comment by zapov** [comment]
It would be helpful if you could provide more info, but you can pair discriminator with name to have a value for each subtype; which is I'm assuming what you want.

DSL-JSON does not support adapters for such configuration at the moment, although it should be relatively straightforward to implement it.

**Comment by sky-pedrosaraiva** [comment]
Yes, you are right. I tried that and now gives me this error:
`com.dslplatform.json.ParsingException: Expecting "type" attribute as first element of mixin com.test.client.models.Dummy. Found: 'items' at position: 21, following: `"group":{"dummy":[{"`, before: `items":[{"classifica``

Unfortunatly I can't assure that type will be the first item because it comes from external API. Is there a workaround for this?

**Comment by zapov** [comment]
Not really. The only workaround is you implement your own converter with this logic.
But to handle this in a general case it is not really performant and thus there is no plan to change how this works.

**Comment by sky-pedrosaraiva** [comment]
I have another issue related to @JsonConverter.
My project consists in multiple gradle modules (i.e feat1 and feat2).
In feat1 I declare a converter like this and it works fine within classes annotated with @CompiledJson in feat1:

```
@JsonConverter(target = Instant::class)
public object InstantConverter {
    public fun read(reader: JsonReader<*>): Instant? {
        return if (reader.wasNull()) null else Instant.parse(reader.readSimpleString())
    }

    public fun write(writer: JsonWriter, value: Instant?) {
        if (value == null) {
            writer.writeNull()
        } else {
            writer.writeString(value.toString())
        }
    }
}
```
But if in feat2 there's a @CompiledJson with an Instant field, it fails to compile because of missing converter.
I've tried to annotate that field with @JsAttribute(converter = ...) but it still fails to compile.

Any ideas?

**Comment by zapov** [comment]
The only thing which comes to mind is some problem with incremental compilation.
So try to use whole project compilation and see if that helps.

**Comment by sky-pedrosaraiva** [comment]
Can't get it to work.
I made a simple app with two modules were this issue is present:
`https://github.com/sky-pedrosaraiva/dsl-json-test/tree/main`

The converter is here:
`https://github.com/sky-pedrosaraiva/dsl-json-test/blob/b039972b93676ab895f7ce9b4d2faa740f0e4736/app/lib/src/main/java/com/psar/lib/MyClass.kt#L9C11-L9C11`

**Comment by zapov** [comment]
I tried looking into your project but compilation fails with
```
A problem occurred configuring root project 'test'.
> Could not resolve all files for configuration ':classpath'.
   > Could not resolve com.android.tools.build:gradle:8.1.0-rc01.
     Required by:
         project : > com.android.application:com.android.application.gradle.plugin:8.1.0-rc01
```
I dont have time to fix it or debug why its not compiling out of the box

**Comment by sky-pedrosaraiva** [comment]
Please try this one:
[Repo](https://github.com/sky-pedrosaraiva/dsl-json-test2)

**Comment by zapov** [comment]
Still doesn't work.
I had to change all kind of stuff to get it working but in the end it fails with

```
This version (1.3.2) of the Compose Compiler requires Kotlin version 1.7.20 but you appear to be using Kotlin version 1.8.20 which is not known to be compatible.  Please fix your configuration (or `suppressKotlinVersionCompatibilityCheck` but don't say I didn't warn you!).
```

Anyway... from what I've seen I dont see why you expect this to work at all. There is no dependency between the projects (lib1 and lib2)
It seems to me it should work as long as lib2 can see this setup from lib1 (or has its own).
